ethereum/contracts/BeefyLightClient.sol

```diff
@@ -383,18 +368,30 @@ contract BeefyLightClient {
      * @param nextAuthoritySetLen The number of validators in the next authority set
      * @param nextAuthoritySetRoot The merkle root of the merkle tree of the next validators
      */
-    function applyValidatorSetChanges(
-        uint64 nextAuthoritySetId,
-        uint32 nextAuthoritySetLen,
-        bytes32 nextAuthoritySetRoot
+    function updateValidatorSet(
+        bytes32 root,
+        BeefyMMRLeaf calldata leaf,
+        SimplifiedMMRProof calldata proof
     ) internal {
-        if (nextAuthoritySetId != validatorRegistry.id()) {
-            validatorRegistry.update(
-                nextAuthoritySetRoot,
-                nextAuthoritySetLen,
-                nextAuthoritySetId
-            );
+
+        // Don't process older leafs
+        if (leaf.nextAuthoritySetId <= validatorRegistry.id()) {
+            return;
         }
+
```

Comment:
so, beefy is changing to increase the id every time now right?

is there any cheap way we can check if the validator registry has actually changed or not so that we can skip the more expensive inclusion proof check if the validator set is all the same?

maybe if nextAuthoritySetRoot hasn't changed we could skip the verification, but i guess that opens up an attack where an incumbent validator set can frontrun a new validator set and get the bridge to fork.

we'd need beefy to provide us with a field that does not increment that allows us to skip this

Reply:
> so, beefy is changing to increase the id every time now right?

Not really. Its just the following that is happening (according to Adrian):
1. The beefy validator set ID will increment on every new session, regardless of whether the validator root changes or not.
2. The session length on test networks is roughly 1 minutes. Not 4 hours like in production Polkadot

Change (1) is new and is part of Polkadot 0.9.17 / Lean Beefy. It's meant to ensure that each session has a unique validator set id, for high-level economic reasons (according to Adrian).

So the effect of (1) and (2) combined is that we see validator set ids incremented every minute (or roughly every 2 commitments on our E2E stack).

Don't think we need to change anything to improve performance. On production Polkadot and Kusama the IDs will only change every hour or more.



 On test polkadot networks such as our E2E stack its changing roughly 10 blocks to help tease out bugs. Is a new thing in Polkadot 0.9.17 / Lean BEEFY. However on production Polkadot the validator set id will only change every 2400 blocks as usual (regardless of the validator set root).



The validator set is still only changing every 2400 blocks on Polkadot. However